lib/path/default.nix

```
@@ -149,6 +174,104 @@ in /* No rec! Add dependencies on this file at the top. */ {
           ${subpathInvalidReason subpath}'';
     path + ("/" + subpath);
 
+  /*
+  Whether the second path is a prefix of the first path, or equal to it.
+  Throws an error if the paths don't share the same filesystem root.
+
+  Laws:
+
+  - Equivalent to whether some subpath exists that can be appended to the first path to get the second path:
+
+        hasPrefix p q <-> exists s . append p s == q
+
+  - `hasPrefix` is a [non-strict partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Non-strict_partial_order) over the set of all path values
+
+  Type:
+    hasPrefix :: Path -> Path -> Bool
+
+  Example:
+    hasPrefix /foo /foo/bar
+    => true
+    hasPrefix /foo /foo
+    => true
+    hasPrefix /foo/bar /foo
+    => false
+    hasPrefix /. /foo
+    => true
+  */
+  hasPrefix =
+    path1:
+    assert assertMsg
+      (isPath path1)
+      "lib.path.hasPrefix: First argument is of type ${typeOf path1}, but a path was expected";
+    let
+      path1Deconstructed = deconstructPath path1;
+    in
+      path2:
+      assert assertMsg
+        (isPath path2)
+        "lib.path.hasPrefix: Second argument is of type ${typeOf path2}, but a path was expected";
+      let
+        path2Deconstructed = deconstructPath path2;
+      in
+        assert assertMsg
+        (path1Deconstructed.root == path2Deconstructed.root) ''
+          lib.path.hasPrefix: Filesystem roots must be the same for both paths, but paths with different roots were given:
+              first argument: "${toString path1}" (root "${toString path1Deconstructed.root}")
+              second argument: "${toString path2}" (root "${toString path2Deconstructed.root}")'';
```

```suggestion
    path1: assert assertMsg (isPath path1) "lib.path.hasPrefix: First argument is of type ${typeOf path1}, but a path was expected";
    path2: assert assertMsg (isPath path2) "lib.path.hasPrefix: Second argument is of type ${typeOf path2}, but a path was expected";
      let
        path1Deconstructed = deconstructPath path1;
        path2Deconstructed = deconstructPath path2;
      in
        assert assertMsg (path1Deconstructed.root == path2Deconstructed.root) ''
          lib.path.hasPrefix: Filesystem roots must be the same for both paths, but paths with different roots were given:
              first argument: "${toString path1}" with root "${toString path1Deconstructed.root}"
              second argument: "${toString path2}" with root "${toString path2Deconstructed.root}"'';

```

Changes:
 - a slight refactor that saves an Env and is a bit more readable in my opinion.
 - removed parentheses from the error message.
 - added a newline before the actual work starts for readability.

Consider these for `removePrefix` as well.

---

Alright then I guess. ViewPatterns would be nice.

I fell into this rabbit hole by the way, while unaware of the uncached transformation https://github.com/NixOS/nix/issues/8211

It stops just short of ViewPatterns, unless `@=` isn't deduplicated or has a non-deduped variant (though I guess at that point we need keywords instead of operator soup).

---

Subscribed to that issue, though I don't want to get into that rabbit hole myself right now :)

I applied your other suggestion to remove the parenthesis from the error message though